src/oom/oomd-manager.c

```diff
@@ -0,0 +1,267 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include "cgroup-util.h"
+#include "fd-util.h"
+#include "fileio.h"
+#include "oomd-manager.h"
+#include "path-util.h"
+#include "varlink.h"
+
+/* Fill new_h with path and path's children cgroups data. It ignores most errors in order to handle cgroups that may
+ * get cleaned up while we populate the hashmap.
+ * saved_h may be NULL (in which case there was no prior data to reference).
+ *
+ * saved_h and new_h are of the form { key: cgroup paths -> value: OomdCGroupContext } */
+static int recursively_get_cgroup_context(Hashmap *saved_h, Hashmap *new_h, const char *path) {
+        _cleanup_free_ OomdCGroupContext *curr_ctx = NULL;
+        _cleanup_closedir_ DIR *d = NULL;
+        _cleanup_free_ char *key = NULL;
+        OomdCGroupContext *saved_ctx;
+        char *subpath;
+        int r;
+
+        assert(new_h);
+        assert(path);
+
+        r = cg_enumerate_subgroups(SYSTEMD_CGROUP_CONTROLLER, path, &d);
+        if (r < 0)
+                return r;
+
+        while ((r = cg_read_subgroup(d, &subpath)) > 0) {
+                _cleanup_free_ char *p = NULL;
+
+                p = path_join(empty_to_root(path), subpath);
+                free(subpath);
+                if (!p)
+                        return log_oom();
+
+                r = recursively_get_cgroup_context(saved_h, new_h, p);
+                if (r == -ENOMEM)
+                        return log_oom();
+        }
+
+        curr_ctx = new(OomdCGroupContext, 1);
+        if (!curr_ctx)
+                return -ENOMEM;
```

Comment:
some oom errors here fail with logging and some without. each function in systemd should either log about all errors or about none (in which case the caller should do that). library code should usually do the latter, main program code the former. I figure this is main program code, hence this should log i guess